### PR TITLE
[ot_certs] Add support for validity dates

### PR DESCRIPTION
**Pull request by pamaury**
Only review the last commit, the rest is from #22284

Tagging @jwnrt and @nbdd0121 for unsafe review.

### CODE REVIEW

File: sw/device/silicon_creator/lib/cert/cdi_1.hjson [hunk 1]
```diff
@@ -63,6 +63,13 @@
         subject: {
             serial_number: { var: "owner_pub_key_id", convert: "lowercase-hex" },
         },
+        // From the Open Profile for DICE specification:
+        // https://pigweed.googlesource.com/open-dice/+/refs/heads/main/docs/specification.md#x_509-cdi-certificates
+        // The certificate expiration time is set to a fixed date in the future.
+        // The "not before" date is chosen to be in the past since the device does not
+        // have a reliable way to get the time.
+        not_before: "20230101000000Z",
```
Comment (timothytrippel) [inline sw/device/silicon_creator/lib/cert/cdi_1.hjson:71]: Just curious: how did you select the "not_before" value? Looks like the [open profile](https://pigweed.googlesource.com/open-dice/+/refs/heads/main/docs/specification.md?pli=1#x_509-cdi-certificates) had a different value (which corresponded to the date of publication of the TCG DICE spec. Not that it matters, just was curious :)

**Review by timothytrippel — APPROVED** [review]

File: sw/device/silicon_creator/lib/cert/cdi_1.hjson (hunk 1 above)
Comment (pamaury) [reply sw/device/silicon_creator/lib/cert/cdi_1.hjson:71]: The DICE specification said any date in the past is acceptable so I just took a random date. But it's true that the publication date of the spec makes the most sense ^^ I am happy to change if you think that's worth it, it's now it's trivial to change in the hjson template.

File: sw/host/ot_certs/src/x509.rs [hunk 2]
```diff
@@ -75,6 +76,35 @@ fn asn1octets_to_vec(s: &Asn1OctetStringRef) -> Value<Vec<u8>> {
     Value::literal(s.as_slice().to_vec())
 }
 
+fn asn1time_to_string(time: &Asn1TimeRef) -> Result<Value<String>> {
+    // SAFETY: OpenSSL guarantees that an ASN1_TIME is in fact just a typedef for ASN1_STRING
```
Comment (nbdd0121) [inline sw/host/ot_certs/src/x509.rs:80]: nit, as this does not have an unsafe block:
```suggestion
    // OpenSSL guarantees that an ASN1_TIME is in fact just a typedef for ASN1_STRING
```

File: sw/host/ot_certs/src/x509.rs [hunk 3]
```diff
@@ -75,6 +76,35 @@ fn asn1octets_to_vec(s: &Asn1OctetStringRef) -> Value<Vec<u8>> {
...
+        bail!("time uses type {time_type} but only GeneralizedTime is supported")
+    }
+    // SAFETY: the above pointer is guaranteed to be valid since `time` is valid reference.
+    // From the OpenSSL documentation: ASN1_STRING_get0_data() returns an internal pointer to
+    // the data of x. Since this is an internal pointer it should not be freed or modified in any way.
+    let time_str = unsafe {
+        let len = openssl_sys::ASN1_STRING_length(time_str);
+        let data = openssl_sys::ASN1_STRING_get0_data(time_str);
+        std::slice::from_raw_parts(data, len as usize)
+    };
```
Comment (nbdd0121) [inline sw/host/ot_certs/src/x509.rs:100]: How about just use
```rust
let time_str = unsafe { Asn1StringRef::from_ptr(time_str) }.as_str();
```

Comment (pamaury) [reply sw/host/ot_certs/src/x509.rs:100]: I should I thought about that ^^ Thanks for the suggestion.

File: sw/device/silicon_creator/lib/cert/cdi_1.hjson (hunk 1 above)
Comment (timothytrippel) [reply sw/device/silicon_creator/lib/cert/cdi_1.hjson]: no need to change, was just curious

**Pull request merged**